File: python/leetcodehard/wildcard_matching.py
"""
https://leetcode.com/problems/wildcard-matching/

status - completed
"""

from base_logger.logging_event import create_logger


class Solution:

    log = create_logger(__name__)

    def isMatch(self, s, p):
        rows = len(s)
        cols = len(p)
        dp = [[False] * (cols + 1) for _ in range(rows + 1)]
        dp[0][0] = True

        for j in range(1, cols + 1):
            if p[j - 1] == "*":
                dp[0][j] = dp[0][j - 1]

        for i in range(1, rows + 1):
            for j in range(1, cols + 1):
                if p[j - 1] == "*":
                    dp[i][j] = dp[i - 1][j] or dp[i][j - 1]
                elif p[j - 1] == "?" or p[j - 1] == s[i - 1]:
                    dp[i][j] = dp[i - 1][j - 1]

        return dp[rows][cols]


if __name__ == "__main__":
    test_input_1 = ("aa", "*")
    output = Solution().isMatch(*test_input_1)
    Solution().log.info("Output is %s", output)

    test_input_2 = ("cb", "?a")
    output = Solution().isMatch(*test_input_2)
    Solution().log.info("Output is %s", output)

File: typescript/leetcodehard/wildcardMatching.ts
/*
https://leetcode.com/problems/wildcard-matching/

status - completed
*/

import { logp } from "../utils/logger";

function isMatch(s: string, p: string): boolean {
  const m = s.length;
  const n = p.length;
  const dp: boolean[][] = Array.from({ length: m + 1 }, () => new Array(n + 1).fill(false));
  dp[0][0] = true;

  for (let j = 1; j <= n; j++) {
    if (p[j - 1] === "*") {
      dp[0][j] = dp[0][j - 1];
    }
  }

  for (let i = 1; i <= m; i++) {
    for (let j = 1; j <= n; j++) {
      if (p[j - 1] === "*") {
        dp[i][j] = dp[i - 1][j] || dp[i][j - 1];
      } else if (p[j - 1] === "?" || p[j - 1] === s[i - 1]) {
        dp[i][j] = dp[i - 1][j - 1];
      }
    }
  }

  return dp[m][n];
}

logp(`input s="aa" p="a" expected false actual ${isMatch("aa", "a")}`);
logp(`input s="adceb" p="*a*b" expected true actual ${isMatch("adceb", "*a*b")}`);
logp(`input s="cb" p="?a" expected false actual ${isMatch("cb", "?a")}`);
